### PR TITLE
BorderBoxControl: migrate tests to TypeScript, remove `act()` call

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Part of #35744 

Refactor `BorderBoxControl`'s tests to TypeScript (the rest of the component is already typed)

Remove unnecessary `act()` call in the process

## Why?

Part of the TypeScript migration that we're doing in the `@wordpress/components` package, see #35744 for more info

## Testing Instructions

No runtime changes. Make sure that the project compiles, and that tests make sense

### CODE REVIEW

File: packages/components/src/border-box-control/test/index.tsx [hunk 1]
```diff
@@ -189,7 +188,14 @@ describe( 'BorderBoxControl', () => {
...
 			await user.click( colorButton );
-			await act( () => Promise.resolve() );
+
+			await waitFor( () =>
+				expect(
+					screen.getByRole( 'button', {
+						name: 'Custom color picker.',
+					} )
+				).toBeVisible()
+			);
```
Comment (ciampo) [inline packages/components/src/border-box-control/test/index.tsx:207]: @tyxla  this change is actually triggering the well-know "An update to %s inside a test was not wrapped in act(...)." warning.

Any ideas on how to fix it? I tried following your approach from other PRs but couldn't make it work.

Thank you!

Comment (ciampo) [reply packages/components/src/border-box-control/test/index.tsx:207]: I realised that I had forgotten to stop using fake timers — did that in https://github.com/WordPress/gutenberg/pull/47755/commits/ca41ab04bde65239566e0836fcb031b83b9d0dc8. It should hopefully fix the issue 🤞

Comment (ciampo) [reply packages/components/src/border-box-control/test/index.tsx:207]: It still looks like there may be timeout issues — I'd appreciate your help in understanding how to fix those errors. Thank you!

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for finishing off the conversion of the BorderBoxControl to TypeScript @ciampo 👍 

The changes look good.

✅ No typing errors
✅ Unit tests are passing locally.
✅ Updates look good. No runtime changes.

Regarding the remaining timeout error in the failed e2e, that looks related to a [different test](https://github.com/WordPress/gutenberg/pull/47755/files#diff-3826762a1aa0d8439b458732668b3cbc4a9b36f7b246caf716b507914420597cR283) than the one commented on in this PR. I couldn't, however, replicate the problem locally. 🤔

File: packages/components/src/border-box-control/test/index.tsx (hunk 1 above)
Comment (tyxla) [reply packages/components/src/border-box-control/test/index.tsx:207]: Sure thing! I've filed a commit and will follow up with a comment to explain.

Comment (tyxla) [reply packages/components/src/border-box-control/test/index.tsx:207]: I'm actually unable to repro this at all. Perhaps using real timers fixed it?

File: packages/components/src/border-box-control/test/index.tsx [hunk 2]
```diff
@@ -260,27 +273,34 @@ describe( 'BorderBoxControl', () => {
...
 
-		it( 'should omit style options when requested', async () => {
-			const user = userEvent.setup( {
-				advanceTimers: jest.advanceTimersByTime,
-			} );
+		// We're expecting to have 4 color buttons by default.
+		const colorButtonIndexes = [ ...Array( 4 ).keys() ];
 
-			render( <BorderBoxControl { ...props } enableStyle={ false } /> );
+		it.each( colorButtonIndexes )(
```
Comment (tyxla) [inline packages/components/src/border-box-control/test/index.tsx:279]: I've noticed that this test times out on CI (not locally though) so in https://github.com/WordPress/gutenberg/pull/47755/commits/bde3e094416d2ddb0683eb5b16c923fde117090d I've split it to separate tests that target each button separately. This should distribute the effort and make tests run within the 5s limit.

File: packages/components/src/border-box-control/test/index.tsx [hunk 3]
```diff
@@ -260,27 +273,34 @@ describe( 'BorderBoxControl', () => {
...
+
+				const colorButtons =
+					screen.getAllByLabelText( colorPickerRegex );
 
-			const colorButtons = screen.getAllByLabelText( colorPickerRegex );
+				await user.click( colorButtons[ colorButtonIndex ] );
 
-			function assertStyleOptionsMissing() {
 				const styleLabel = screen.queryByText( 'Style' );
 				const solidButton = screen.queryByRole( 'button', {
```
Comment (tyxla) [inline packages/components/src/border-box-control/test/index.tsx]: We could be able to shorten this to a single query, where:

```
const solidButton = screen.queryByRole( 'button', {
	name: 'Solid',
} );
const dashedButton = screen.queryByRole( 'button', {
	name: 'Dashed',
} );
const dottedButton = screen.queryByRole( 'button', {
	name: 'Dotted',
} );

expect( solidButton ).not.toBeInTheDocument();
expect( dashedButton ).not.toBeInTheDocument();
expect( dottedButton ).not.toBeInTheDocument();

```

to:

```
const styleButton = screen.queryByRole( 'button', {
	name: /(Solid)|(Dashed)|(Dotted)/,
} );

expect( styleButton ).not.toBeInTheDocument();
```

This is totally optional since it's a suggestion on pre-existing code.

**Review by tyxla — APPROVED** [review]
I've added a commit to fix the test timeout and left some thoughts, but this looks great IMHO 👍 🚀 Nice work

File: packages/components/src/border-box-control/test/index.tsx (hunk 1 above)
Comment (ciampo) [reply packages/components/src/border-box-control/test/index.tsx:207]: Real timers fixed it locally, but your commit https://github.com/WordPress/gutenberg/pull/47755/commits/bde3e094416d2ddb0683eb5b16c923fde117090d fixed on CI as well. Thank you 🙏

File: packages/components/src/border-box-control/test/index.tsx (hunk 2 above)
Comment (ciampo) [reply packages/components/src/border-box-control/test/index.tsx:279]: That fixed it, thank you!

File: packages/components/src/border-box-control/test/index.tsx (hunk 3 above)
Comment (ciampo) [reply packages/components/src/border-box-control/test/index.tsx]: Clever! Applied in https://github.com/WordPress/gutenberg/pull/47755/commits/d7b188b251743e11181e81cf551442f57f5c9c82

Comment (tyxla) [reply packages/components/src/border-box-control/test/index.tsx]: Nice. You might also want to do it for the other `should omit style options when requested` test that has the same assertions.
